ntp-proto/src/system.rs

```diff
@@ -0,0 +1,127 @@
+use crate::filter::MAX_STRATUM;
+use crate::{
+    filter::{clock_combine, clock_select, multiply_by_phi, Peer},
+    packet::NtpLeapIndicator,
+    NtpDuration, NtpTimestamp,
+};
+
+// NOTE stored in exponent format in the spec
+const MIN_POLL: NtpDuration = NtpDuration::from_bits_short((16u32 << 16u32).to_be_bytes()); // 16 seconds
+
+struct System {
+    time: NtpTimestamp,
+    jitter: NtpDuration,
+    offset: NtpDuration,
+
+    leap: NtpLeapIndicator,
+    stratum: u8,
+
+    poll: NtpDuration,
+
+    root_delay: NtpDuration,
+    root_dispersion: NtpDuration,
+
+    reference_id: u32,
+    reference_timestamp: NtpTimestamp,
+}
+
+impl System {
+    #[allow(dead_code)]
+    fn clock_update(
+        &mut self,
+        peers: &[Peer],
+        local_clock_time: NtpTimestamp,
+        system_poll: NtpDuration,
+    ) {
+        let survivors = match clock_select(peers, local_clock_time, system_poll) {
+            None => return,
+            Some(v) => v,
+        };
+
+        // TODO, in the implementation (but not the spec from what I've seen)
+        // the current system peer is kept if it is a survivor and it's stratum
+        // is equal to the best (i.e. index 0) of the survivors
+        let p = survivors[0].peer;
+
+        if self.time >= p.time {
+            return;
+        }
+
+        let combined = clock_combine(&survivors, local_clock_time);
+
+        self.time = p.time;
+        self.offset = combined.offset;
+        self.jitter = combined.jitter;
+
+        match local_clock(p, self.offset) {
+            LocalClockCode::Panic => {
+                // The offset is too large and probably bogus.  Complain to the
+                // system log and order the operator to set the clock manually
+                // within PANIC range.  The reference implementation includes a
+                // command line option to disable this check and to change the
+                // panic threshold from the default 1000 s as required.
+                panic!("fatal error");
```

Comment:
the skeleton just does `exit(0)` here. I think panicking is the right option here (we already panic in some cases where the state is just invalid)